### PR TITLE
Add level 1 routine ddot

### LINKED ISSUE #262: RFC: Add BLAS Level 1 ddot routine
## Checklist

> Please ensure the following tasks are completed before submitting a feature request.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] The issue name begins with `RFC:`.

## Description

> Description of the feature request.

This RFC proposes implementing the dot (or inner) product routine for a double-precision floating-pint arrays as defined in [BLAS Level 1][blas-level-1] routines. Specifically adding `@stdlib/blas/base/ddot` is proposed.

## Related Issues

> Does this feature request have any related issues?

Related issues #261 

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this feature request? This may include screenshots, references, sample output, and/or implementation notes.

No.

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[blas-level-1]: http://www.netlib.org/blas/#_level_1

### PR DESCRIPTION
**Pull request by splrk**
Resolves #262.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request implements the BLAS level 1 routine `@stdlib/blas/base/ddot`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #262.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

### CODE REVIEW

File: lib/node_modules/@stdlib/blas/base/ddot/docs/repl.txt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    y: Float64Array
+        Second input array.
+
+    strideY: integer
+        Index increment for `y`.
+
+    Returns
+    -------
+    dot: float
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/docs/repl.txt]:
```suggestion
    dot: number
```

File: lib/node_modules/@stdlib/blas/base/ddot/docs/repl.txt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    strideY: integer
+        Index increment for `y`.
+
+    offsetY: integer
+        Starting index for `y`.
+
+    Returns
+    -------
+    dot: float
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/docs/repl.txt]:
```suggestion
    dot: number
```

File: lib/node_modules/@stdlib/blas/base/ddot/lib/ddot.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		if ( m > 0 ) {
+			for ( i = 0; i < m; i++ ) {
+				dot += x[ i ] * y[ i ];
+			}
+		}
+		if ( N < M ) {
+			return dot;
+		}
+		for ( i = m; i < N; i += M ) {
+			dot = dot + ( x[ i ] * y[ i ] ) + ( x[ i+1 ] * y[ i+1 ] ) + ( x[ i+2 ] * y[ i+2 ] ) + ( x[ i+3 ] * y[ i+3 ] ) + ( x[ i+4 ] * y[ i+4 ] ); // eslint-disable-line max-len
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/lib/ddot.js]:
```suggestion
			dot += ( x[ i ] * y[ i ] ) + ( x[ i+1 ] * y[ i+1 ] ) + ( x[ i+2 ] * y[ i+2 ] ) + ( x[ i+3 ] * y[ i+3 ] ) + ( x[ i+4 ] * y[ i+4 ] ); // eslint-disable-line max-len
```

File: lib/node_modules/@stdlib/blas/base/ddot/lib/ddot.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	} else {
+		ix = 0;
+	}
+	if ( strideY < 0 ) {
+		iy = ( 1-N ) * strideY;
+	} else {
+		iy = 0;
+	}
+	for ( i = 0; i < N; i++ ) {
+		dot = dot + ( x[ ix ] * y[ iy ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/lib/ddot.js]:
```suggestion
		dot += x[ ix ] * y[ iy ];
```

File: lib/node_modules/@stdlib/blas/base/ddot/lib/ndarray.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	iy = offsetY;
+
+	// Use unrolled loops if both strides are equal to `1`...
+	if ( strideX === 1 && strideY === 1 ) {
+		m = N % M;
+
+		// If we have a remainder, run a clean-up loop...
+		if ( m > 0 ) {
+			for ( i = 0; i < m; i++ ) {
+				dot = dot + ( x[ ix ] * y[ iy ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/lib/ndarray.js]:
```suggestion
				dot += x[ ix ] * y[ iy ];
```

File: lib/node_modules/@stdlib/blas/base/ddot/lib/ndarray.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+				ix += 1;
+				iy += 1;
+			}
+		}
+		if ( N < M ) {
+			return dot;
+		}
+		for ( i = m; i < N; i += M ) {
+			dot = dot + ( x[ ix ] * y[ iy ] ) + ( x[ ix+1 ] * y[ iy+1 ] ) + ( x[ ix+2 ] * y[ iy+2 ] ) + ( x[ ix+3 ] * y[ iy+3 ] ) + ( x[ ix+4 ] * y[ iy+4 ] ); // eslint-disable-line max-len
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/lib/ndarray.js]:
```suggestion
			dot += ( x[ ix ] * y[ iy ] ) + ( x[ ix+1 ] * y[ iy+1 ] ) + ( x[ ix+2 ] * y[ iy+2 ] ) + ( x[ ix+3 ] * y[ iy+3 ] ) + ( x[ ix+4 ] * y[ iy+4 ] ); // eslint-disable-line max-len
```

File: lib/node_modules/@stdlib/blas/base/ddot/lib/ndarray.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+			ix += M;
+			iy += M;
+		}
+		return dot;
+	}
+	for ( i = 0; i < N; i++ ) {
+		dot = dot + ( x[ ix ] * y[ iy ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/lib/ndarray.js]:
```suggestion
		dot += x[ ix ] * y[ iy ];
```

File: lib/node_modules/@stdlib/blas/base/ddot/src/addon.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/addon.cpp]: You can remove this file, as I need to revisit and refactor due to deprecated V8 APIs. :(

File: lib/node_modules/@stdlib/blas/base/ddot/src/ddot.f [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+! > The reference BLAS is a freely-available software package. It is available from netlib via anonymous ftp and the World Wide Web. Thus, it can be included in commercial software packages (and has been). We only ask that proper credit be given to the authors.
+! >
+! > Like all software, it is copyrighted. It is not trademarked, but we do ask the following:
+! >
+! > * If you modify the source for these routines we ask that you change the name of the routine and comment the changes made to the original.
+! >
+! > * We will gladly answer any questions regarding the software. If a modification is done, however, it is the responsibility of the person who modified the routine to provide support.
+!
+! @param {integer} N - number of values over which to compute the dot product
+! @param {Array<double precision>} sx - first array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/ddot.f]:
```suggestion
! @param {Array<double>} sx - first array
```

File: lib/node_modules/@stdlib/blas/base/ddot/src/ddot.f [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+! @param {integer} strideX - `sx` stride length
+! @param {Array<double precision>} sy - second array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/ddot.f]:
```suggestion
! @param {Array<double>} sy - second array
```

File: lib/node_modules/@stdlib/blas/base/ddot/src/ddot.f [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+! @param {integer} strideY - `sy` stride length
+! @returns {double precision} the dot product of `sx` and `sy`
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/ddot.f]:
```suggestion
! @returns {double} the dot product of `sx` and `sy`
```

File: lib/node_modules/@stdlib/blas/base/ddot/src/ddotsub.f [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+!<
+
+!> Wraps ddot as a subroutine.
+
+! @param {integer} N - number of values over which to compute the dot product
+! @param {Array<double precision>} sx - first array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/ddotsub.f]:
```suggestion
! @param {Array<double>} sx - first array
```

File: lib/node_modules/@stdlib/blas/base/ddot/src/ddotsub.f [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+! @param {integer} strideX - `sx` stride length
+! @param {Array<double precision>} sy - second array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/ddotsub.f]:
```suggestion
! @param {Array<double>} sy - second array
```

File: lib/node_modules/@stdlib/blas/base/ddot/src/ddotsub.f [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+! @param {integer} strideY - `sy` stride length
+! @param {double precision} dot - output variable reference
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/ddot/src/ddotsub.f]:
```suggestion
! @param {double} dot - output variable reference
```

File: lib/node_modules/@stdlib/blas/base/sdot/src/addon.cpp [hunk 15]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/sdot/src/addon.cpp]: This file can be removed due to reasons mentioned above.

File: lib/node_modules/@stdlib/namespace/lib/namespace/b.js [hunk 16]
```diff
@@ -859,6 +859,18 @@ ns.push({
 	]
 });
 
+ns.push({
+	'alias': 'base.ddot',
+	'path': '@stdlib/blas/base/ddot',
+	'value': require( '@stdlib/blas/base/ddot' ),
+	'type': 'Function',
+	'related': [
+		'@stdlib/blas/base/sdot',
```
Comment (kgryte) [inline lib/node_modules/@stdlib/namespace/lib/namespace/b.js]: Need to reorder related to be in alphabetical order.

File: lib/node_modules/@stdlib/namespace/alias2pkg/data/data.csv [hunk 17]
```diff
@@ -109,6 +109,7 @@ base.csub,"@stdlib/math/base/complex/subtract"
 base.dasum,"@stdlib/blas/base/dasum"
 base.daxpy,"@stdlib/blas/base/daxpy"
 base.dcopy,"@stdlib/blas/base/dcopy"
+base.ddot,"@stdlib/blas/base/ddot"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/namespace/alias2pkg/data/data.csv:112]: Thanks for updating these files. Just out of curiosity, how are you updating these database files? I am hoping not manually. :)

**Review by kgryte — CHANGES_REQUESTED** [review]
@splrk Really awesome work! Only a few minor nitpicks. Once resolved, this should be ready for merge! Thanks for working on this!

File: lib/node_modules/@stdlib/namespace/alias2pkg/data/data.csv (hunk 17 above)
Comment (splrk) [reply lib/node_modules/@stdlib/namespace/alias2pkg/data/data.csv:112]: What kind of programmer would I be if I did things manually? :-D

I am using `make repl-docs` after I edit `@stdlib/namespace/lib/namespace/b.js`.  I was a little frustrated that I couldn't see my `repl.txt` docs in the REPL, so I went fishing in the repo.  I first found all the `build.js` files in `namespace` which led me to `make repl-docs`.

**Comment by splrk** [comment]
Sorry for the late reply on this one.   The weekend got busy.

File: lib/node_modules/@stdlib/namespace/alias2pkg/data/data.csv (hunk 17 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/namespace/alias2pkg/data/data.csv:112]: Nice! Excellent sleuthing to discover `make repl-docs`! :)

**Review by kgryte — APPROVED** [review]
@splrk Excellent work! Thanks for working on this!

**Pull request merged**